src/connector/src/source/mqtt/enumerator/mod.rs

```diff
@@ -0,0 +1,144 @@
+            topics.insert(topic.clone());
+        }
+
+        client
+            .subscribe(topic.clone(), rumqttc::v5::mqttbytes::QoS::AtMostOnce)
+            .await?;
+
+        let cloned_client = client.clone();
+
+        let topics = Arc::new(RwLock::new(topics));
+
+        let connected = Arc::new(AtomicBool::new(false));
+        let connected_clone = connected.clone();
+
+        let stopped = Arc::new(AtomicBool::new(false));
+        let stopped_clone = stopped.clone();
+
+        let topics_clone = topics.clone();
+        tokio::spawn(async move {
+            while !stopped_clone.load(std::sync::atomic::Ordering::Relaxed) {
+                match eventloop.poll().await {
+                    Ok(Event::Outgoing(Outgoing::Subscribe(_))) => {
+                        connected_clone.store(true, std::sync::atomic::Ordering::Relaxed);
+                    }
+                    Ok(Event::Incoming(Incoming::Publish(p))) => {
+                        let topic = String::from_utf8_lossy(&p.topic).to_string();
+                        let exist = {
+                            let topics = topics_clone.read().await;
+                            topics.contains(&topic)
+                        };
+
+                        if !exist {
+                            let mut topics = topics_clone.write().await;
+                            topics.insert(topic);
+                        }
+                    }
+                    Ok(_) => {}
+                    Err(err) => {
+                        if let ConnectionError::Timeout(_) = err {
+                            continue;
+                        }
+                        tracing::error!(
+                            "[Enumerator] Failed to subscribe to topic {}: {}",
+                            topic,
+                            err.as_report(),
+                        );
+                        connected_clone.store(false, std::sync::atomic::Ordering::Relaxed);
+                        cloned_client
+                            .subscribe(topic.clone(), rumqttc::v5::mqttbytes::QoS::AtMostOnce)
+                            .await
+                            .unwrap();
+                    }
+                }
+            }
+        });
+
+        Ok(Self {
+            client,
+            topics,
+            topic: properties.common.topic,
+            connected,
+            stopped,
+        })
+    }
+
+    async fn list_splits(&mut self) -> ConnectorResult<Vec<MqttSplit>> {
+        if !self.connected.load(std::sync::atomic::Ordering::Relaxed) {
+            let start = std::time::Instant::now();
+            loop {
+                if self.connected.load(std::sync::atomic::Ordering::Relaxed) {
+                    break;
+                }
+
+                if start.elapsed().as_secs() > 10 {
+                    bail!("Failed to connect to mqtt broker");
+                }
+
+                tokio::time::sleep(std::time::Duration::from_millis(100)).await;
```

Comment:
we use the func here to detect if there is a split change in the upstream, eg. add or remove. 
usually we request it per sec, it is ok to sleep longer.